### DISCUSSION TITLE
Missing warning for `jj diff` treating positional arguments as paths

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#2495] [Ideas]
I've seen `jj log` warn me about treating a positional argument as a path instead of a revision, but it appears that `jj diff <foo>` doesn't give the same warning, even if `<foo>` doesn't exist. Instead, `jj diff <foo>` just has no output if `<foo>` is a revision, which is _very_ surprising and misleading.

Could `jj diff` have a warning about positional arguments that aren't paths to existing files?

**Comment by ilyagr** [comment]
There is an issue and even an unfinished pull request for this. See https://github.com/martinvonz/jj/issues/505.

**Reply by arxanas** [reply]
Should we even allow positional arguments to mean files by default when it gets this confusing? If we just added `-p`/`--path` and required everyone to use that, it seems like it would fix a big usability problem without seriously harming the user interface.

**Reply by ilyagr** [reply]
This would be less convenient for commands like `amend` and `split`, but we could just make `-p` a noop for them, like `-r` for `new`. I think it only needs to be mandatory for `log`, `diff`, and closely related commands (interdiff, obslog)

**Reply by BatmanAoD** [reply]
FWIW, given that `git diff` requires paths to come after `--`, and otherwise treats positional arguments as revisions, I think the positional arguments for `jj diff` are a huge footgun, and strongly agree with getting rid of them.

**Reply by evanrelf** [reply]
Not true:
```ShellSession
$ git init my-repo
Initialized empty Git repository in /Users/evanrelf/Code/my-repo/.git/
$ cd my-repo/
$ echo "Hello, world!" > README.md
$ echo "(c) Me Forever" > LICENSE
$ git add --intent-to-add .
$ git diff README.md
diff --git a/README.md b/README.md
new file mode 100644
index 0000000..af5626b
--- /dev/null
+++ b/README.md
@@ -0,0 +1 @@
+Hello, world!

```

**Reply by BatmanAoD** [reply]
I'd forgotten that `diff` is against the index by default, not `HEAD`. In any case, I still think this is a footgun.

**Reply by martinvonz** [reply]
One potential source of confusion if we required `-p` for path arguments is that shells often expand globbed paths, so e.g. `jj diff -p foo*` would work iff the glob expands to exactly one path.

**Reply by ilyagr** [reply]
That's a very good point.

I had two out-there ideas for dealing with this problem (assuming we can even get `clap` to do this):

1. Show a warning *unless* paths are separated from the other arguments by `--`. So, `jj diff file` and `jj diff -- file` are equivalent except for the fact that the former prints a warning.
2. Have a `-p/--paths` option, so that all the arguments after that are paths *until the next option*.

In fact, we could do both, so that scripts can use option 1 and support files that start with `-`.

This is pretty weird behavior for a CLI, though. Maybe we'll think of something better.

**Reply by BatmanAoD** [reply]
I personally like the first one a lot and am not immediately off-put by the second!

... I do not expect that sentiment to be universal, though 😅

**Reply by necauqua** [reply]
I honestly am a bit put off by the fact that some commands have positional paths and some revisions; always requiring -p and having revisions where applicable (can't even name a command off the top of my head, the whole point is that we can pass @ OR something else to work as an "index" to basically every command) as positional args sounds like an amazing solution of everything ever to me ;)

**Reply by martinvonz** [reply]
I think hg generally uses positional arguments for paths. I suspect the shell-expansion argument I mentioned above is one reason - perhaps the main reason - for that.

**Reply by necauqua** [reply]
Ooh, I didn't initially register what you've meant

But for that you could accept multiple arguments after `-p`.

In fact, I think the model just clicked in my head nicely:

```
jj subcmd [maaaybe sone other positional arg(s), like in branch set] [rev, defaults to @] -p [paths*] -r [rev after paths] -r [maybe even | together multiple rev args] -p [and multiple path args too]
```

I have a strong feeling that this construct I just wrote is literally universal (for commands that don't need paths you don't have the `-p`)
Note how even the currently-stubbed unnecessary -r will work
It seems very neat

and `-p -- -weird-path*` too :)

**Reply by martinvonz** [reply]
I think you're right. I may just be too colored by my use of hg, where there can be exactly one value associated with each flag (but flags and values can be repeated). I'm pretty sure I've seen that clap supports `-p a b c` syntax, so I guess it can't be that uncommon.

I haven't decided what I think about that yet, but is the suggestion then to not have *any* positional arguments? So even `jj new main` would have to be written `jj new -r main`? Maybe I'll get used to it quickly but it sounds a bit annoying.

**Reply by necauqua** [reply]
you probably missed it, but no, the last (optional) positional arg (before 0 up to tons of optional repeating and interleaving -rs and -ps) is a revision, it's perfect

I think literally `grep -r paths* -e 'have to use -e for the actual pattern now'` is like what I proposed

**Reply by martinvonz** [reply]
Ah, I see. I didn't look very carefully, and maybe I'm also too used to the hg CLI where you can put positional arguments anywhere between flags.

**Reply by necauqua** [reply]
The funny part is that since -p accepts any number of args and -r accepts one arg, it should work to have `-p paths* -r rev some-positional-arg`

I think clap handles that fine 😄

**Reply by arxanas** [reply]
I opened https://github.com/martinvonz/jj/issues/3809 as my preferred solution.
